From 6bad27de4e42800bd41430129881f294bfe30c49 Mon Sep 17 00:00:00 2001
From: Jacob Young <jacobly0@users.noreply.github.com>
Date: Sun, 26 Oct 2025 11:12:23 -0400
Subject: [PATCH 027/133] x86_64: continue hacking around unimplemented linker
 logic
X-Developer-Signature: v=1; a=openpgp-sha256; l=1896; i=xtex@astrafall.org;
 h=from:subject; bh=HI09oKdH9WfV2UdnEJwVJRc2IoSh8XPMhi3WPZUzJAo=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cv8cm/Fhzenkbxs2Sk07XVC2Pu7W7+yEKxc79L44f
 W+T/7U9rWMiC4MYF4OlmCJLkWGDN6tOOr/osnJZmDmsTCBDpEUaGBgYGFgY+HIT80qNdIz0TLUN
 9QwNdYx0jBi4OAVgqn9GMvzhCUo+naHc5T2jevMOx2XHxRJ3la/vzItc4vj/cUHedkdZht/sqW9
 3Zp/UE1Blrz9f/YKJ1fHo4tr7j0rLNm15s1CQ6R03AA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Closes #25666

(cherry picked from commit 52a029e503f1c8180a5479871737f01a3b43d1b5)
---
 src/arch/x86_64/CodeGen.zig |  9 ++++++++-
 test/behavior/memset.zig    | 11 +++++++++++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/arch/x86_64/CodeGen.zig b/src/arch/x86_64/CodeGen.zig
index e1ad9e3d074e..39997de7fbf6 100644
--- a/src/arch/x86_64/CodeGen.zig
+++ b/src/arch/x86_64/CodeGen.zig
@@ -181714,7 +181714,14 @@ fn genSetMem(
         .reg => |base_reg| .{ .register_offset = .{ .reg = base_reg, .off = disp } },
         .frame => |base_frame_index| .{ .lea_frame = .{ .index = base_frame_index, .off = disp } },
         .table, .rip_inst, .lazy_sym => unreachable,
-        .nav => |nav| .{ .lea_nav = nav },
+        .nav => |nav| {
+            // hack around linker relocation bugs
+            const addr_reg = try self.copyToTmpRegister(.usize, .{ .lea_nav = nav });
+            const addr_lock = self.register_manager.lockRegAssumeUnused(addr_reg);
+            defer self.register_manager.unlockReg(addr_lock);
+
+            return self.genSetMem(.{ .reg = addr_reg }, disp, ty, src_mcv, opts);
+        },
         .uav => |uav| .{ .lea_uav = uav },
         .extern_func => |extern_func| .{ .lea_extern_func = extern_func },
     };
diff --git a/test/behavior/memset.zig b/test/behavior/memset.zig
index 1efd90a2c6de..d7ee2cf7c00b 100644
--- a/test/behavior/memset.zig
+++ b/test/behavior/memset.zig
@@ -180,3 +180,14 @@ test "@memset with zero-length array" {
     var array: [0]usize = undefined;
     @memset(&array, 0);
 }
+
+test "@memset a global array" {
+    const S = struct {
+        var buf: [1]u32 = .{123};
+    };
+    try expect(S.buf[0] == 123);
+    @memset(&S.buf, 456);
+    try expect(S.buf[0] == 456);
+    @memset(&S.buf, S.buf[0] + 333);
+    try expect(S.buf[0] == 789);
+}
-- 
2.52.0

